src/substitution_model.cpp

```diff
@@ -8,12 +8,114 @@ std::unique_ptr<SubstitutionModel> SubstitutionModel::OfSpecification(
   if (specification == "JC69") {
     return std::make_unique<JC69Model>();
   }  // else
+  if (specification == "HKY") {
+    return std::make_unique<HKYModel>();
+  }  // else
   if (specification == "GTR") {
     return std::make_unique<GTRModel>();
   }  // else
   Failwith("Substitution model not known: " + specification);
 }
 
+void JC69Model::UpdateEigenDecomposition() {
+  eigenvectors_ << 1.0, 2.0, 0.0, 0.5, 1.0, -2.0, 0.5, 0.0, 1.0, 2.0, 0.0, -0.5, 1.0,
+      -2.0, -0.5, 0.0;
+  inverse_eigenvectors_ << 0.25, 0.25, 0.25, 0.25, 0.125, -0.125, 0.125, -0.125, 0.0,
+      1.0, 0.0, -1.0, 1.0, 0.0, -1.0, 0.0;
+  eigenvalues_ << 0.0, -1.3333333333333333, -1.3333333333333333, -1.3333333333333333;
+}
+
+void JC69Model::UpdateQMatrix() {
+  Q_ << -1.0, 1.0 / 3.0, 1.0 / 3.0, 1.0 / 3.0, 1.0 / 3.0, -1.0, 1.0 / 3.0, 1.0 / 3.0,
+      1.0 / 3.0, 1.0 / 3.0, -1.0, 1.0 / 3.0, 1.0 / 3.0, 1.0 / 3.0, 1.0 / 3.0, -1.0;
+}
+
+void HKYModel::SetParameters(const EigenVectorXdRef param_vector) {
+  GetBlockSpecification().CheckParameterVectorSize(param_vector);
+  rates_ = ExtractSegment(param_vector, rates_key_);
+  frequencies_ = ExtractSegment(param_vector, frequencies_key_);
+  if (fabs(frequencies_.sum() - 1.) >= 0.001) {
+    std::ostringstream oss;
+    std::copy(frequencies_.begin(), frequencies_.end() - 1,
+              std::ostream_iterator<double>(oss, ","));
+    oss << *frequencies_.end();
+    Failwith("HKY frequencies do not sum to 1 +/- 0.001! frequency vector: (" +
+             oss.str() + ")");
+  }
+  Update();
+};
+
+void HKYModel::UpdateQMatrix() {
+  EigenVectorXd rates;
+  rates.resize(6);
+  rates << 1.0, rates_[0], 1.0, 1.0, rates_[0], 1.0;
+  int rate_index = 0;
+  for (int i = 0; i < 4; i++) {
+    for (int j = i + 1; j < 4; j++) {
+      double rate = rates[rate_index];
+      rate_index++;
+      Q_(i, j) = rate * frequencies_[j];
+      Q_(j, i) = rate * frequencies_[i];
+    }
+  }
+  // Set the diagonal entries so the rows sum to one.
+  double total_substitution_rate = 0;
+  for (int i = 0; i < 4; i++) {
+    double row_sum = 0;
+    for (int j = 0; j < 4; j++) {
+      if (i != j) {
+        row_sum += Q_(i, j);
+      }
+    }
+    Q_(i, i) = -row_sum;
+    total_substitution_rate += row_sum * frequencies_[i];
+  }
+  // Rescale matrix for unit substitution rate.
+  Q_ /= total_substitution_rate;
+}
+
+// Analytical eigen decomposition
+void HKYModel::UpdateEigenDecomposition() {
+  double kappa = rates_[0];
```

Comment:
This is all standard, but do you have a reference that someone could use to follow along if they wanted?